vibin/mediaservers/cxnv2.py

```diff
@@ -0,0 +1,589 @@
+        self.artwork = artwork
+        self.id_map = _MediaIdGenerator()
+
+        self.tracks = [self.create_track(result) for result in results
+                       if result.type == "object.item.audioItem.musicTrack"]
+        self.tracks_by_id = {track.id: track for track in self.tracks}
+        self.tracks_by_album_id = _Catalogue._group_tracks_by_album_id(self.tracks)
+        self.track_id_by_resource = {_Catalogue.stabilize_resource_uri(result.resource)
+                                     : self.id_map.get_id(result.path)
+                                     for result in results
+                                     if result.resource and result.type == "object.item.audioItem.musicTrack"}
+
+        self.albums = [self.create_album(result) for result in results
+                       if result.type == "object.container.album.musicAlbum"]
+        self.albums_by_id = {album.id: album for album in self.albums}
+
+        self.artists = [self.create_artist(name) for name in
+                        { result.artist for result in results if result.artist }
+                        ]
+        self.artists_by_id = {artist.id: artist for artist in self.artists}
+        self.artists.sort(key = lambda a: a.title)
+
+    def create_track(self, result: _BrowseResult) -> Track:
+        """Creates a new Track with information from this _BrowseResult"""
+        return Track(
+            id = self.id_map.get_id(result.path),
+            albumId = self.id_map.get_id(result.parent_path),
+            title = result.title,
+            artist = result.artist,
+            album = result.album or "Unknown Album",
+            duration = result.duration,
+            genre = result.genre,
+            album_art_uri = self.artwork.get(result.album_art_uri)
+                            or result.album_art_uri,
+            original_track_number = result.original_track_number,
+        )
+
+    def create_album(self, result: _BrowseResult) -> Album:
+        """Creates a new Album with information from this _BrowseResult."""
+        album_id = self.id_map.get_id(result.path)
+        return Album(
+            id = album_id,
+            title = result.title,
+            artist = result.artist or self._guess_artist(album_id),
+            genre = result.genre or self._guess_genre(album_id),
+            album_art_uri = self.artwork.get(result.album_art_uri)
+                            or result.album_art_uri,
+        )
+
+    def _guess_artist(self, album_id: MediaId):
+        """Infers the artist name from an album's tracks.
+
+        If a single artist is responsible for most of the tracks, assume the
+        album is credited to that artist. Otherwise, use the name `Various
+        Artists`.
+        """
+        tracks = self.tracks_by_album_id[album_id]
+        if not tracks:
+            return "Unknown"
+        (artist, count) = Counter([t.artist for t in tracks]).most_common(1)[0]
```

Comment:
Looks like this could raise `IndexError` if `tracks` is empty. Is that something we should be checking for? (No worries if we can trust `tracks` to be non-zero length).

Reply:
Good catch, although I think you mean `KeyError` if the `album_id` isn't present in the `tracks_by_album_id` dict? I've improved this so that the short-circuit return should handle this case. I've also gone through some other places, e.g. `album(self, id)` now throws `VibinNotFoundError` instead of `KeyError`, which matches the implementation in `asset`.

Reply:
Nice. I think your change should address the question I had; but just to make sure, I was wondering about `most_common(1)[0]` throwing if tracks was `[]`:

```python
>>> from collections import Counter
>>> tracks = []
>>> Counter([t.artist for t in []]).most_common(1)[0]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
IndexError: list index out of range
```
I think your modification will also handle this potential issue though, as `[]` is Falsey. So I consider this still resolved unless I'm missing something.